lib/puppet/util/windows/registry.rb

```diff
@@ -30,51 +32,276 @@ def open(name, path, mode = KEY_READ | KEY64, &block)
+
+          subkey_length = subkey_length_ptr.read_dword
+          subkey = subkey_ptr.read_wide_string(subkey_length, Encoding::UTF_8)
+
+          type, data = read(hkey, subkey_ptr)
         end
+      end
+
+      [subkey, type, data]
+    end
+
+    def reg_query_info_key_max_lengths(hkey)
+      result = nil
+
+      FFI::MemoryPointer.new(:dword) do |max_subkey_name_length_ptr|
+        FFI::MemoryPointer.new(:dword) do |max_value_name_length_ptr|
+
+          status = RegQueryInfoKeyW(hkey,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            max_subkey_name_length_ptr, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL, max_value_name_length_ptr,
+            FFI::MemoryPointer::NULL, FFI::MemoryPointer::NULL,
+            FFI::MemoryPointer::NULL
+          )
+
+          if status != FFI::ERROR_SUCCESS
+            raise Puppet::Util::Windows::Error.new('Failed to query registry key info')
+          end
 
-        str.force_encoding(@encoding)
+          result = [
+            # Unicode characters *not* including trailing NULL
+            max_subkey_name_length_ptr.read_dword + 1,
+            max_value_name_length_ptr.read_dword + 1,
+          ]
+        end
       end
-    else
-      def force_encoding(str, enc)
+
+      result
+    end
+
+    # Read a registry value named name and return array of
+    # [ type, data ].
+    # When name is nil, the `default' value is read.
+    # type is value type. (see Win32::Registry::Constants module)
+    # data is value data, its class is:
+    # :REG_SZ, REG_EXPAND_SZ
+    #    String
+    # :REG_MULTI_SZ
+    #    Array of String
+    # :REG_DWORD, REG_DWORD_BIG_ENDIAN, REG_QWORD
+    #    Integer
+    # :REG_BINARY
+    #    String (contains binary data)
+    #
+    # When rtype is specified, the value type must be included by
+    # rtype array, or TypeError is raised.
+    def read(hkey, name_ptr, *rtype)
+      result = nil
+
+      query_value_ex(hkey, name_ptr) do |type, data_ptr, byte_length|
+        unless rtype.empty? or rtype.include?(type)
+          raise TypeError, "Type mismatch (expect #{rtype.inspect} but #{type} present)"
+        end
+
+        string_length = 0
+        # buffer is raw bytes, *not* chars - less a NULL terminator
+        string_length = (byte_length / FFI.type_size(:wchar)) - 1 if byte_length > 0
+
+        case type
+        when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
+          result = [ type, data_ptr.read_wide_string(string_length, Encoding::UTF_8) ]
+        when Win32::Registry::REG_MULTI_SZ
+          result = [ type, data_ptr.read_wide_string(string_length, Encoding::UTF_8).split(/\0/) ]
+        when Win32::Registry::REG_BINARY
+          result = [ type, data.read_bytes(0, byte_length) ]
+        when Win32::Registry::REG_DWORD, Win32::Registry::REG_DWORD_BIG_ENDIAN
+          result = [ type, data_ptr.read_dword ]
```

Comment:
Odd.  I had verified this in the docs earlier, but now I've found conflicting docs (which seem to correlate what Ruby has done).

Or I could have just been really tired and misread.  I will take a look at the headers to verify the definitions, thanks.